scionlab/forms/attachment_link_form.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2018 ETH Zurich
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import ipaddress
+
+from django import forms
+from django.conf import settings
+from django.forms import BaseModelFormSet
+from django.core.exceptions import ValidationError
+
+from crispy_forms.helper import FormHelper
+from crispy_forms.layout import Layout, Row, Column, Div, HTML
+from crispy_forms.bootstrap import AppendedText
+
+from scionlab.defines import MAX_PORT
+from scionlab.models.core import Link
+from scionlab.models.user_as import AttachmentPoint, AttachmentConf, UserAS
+from scionlab.util.portmap import PortMap
+
+
+class AttachmentLinksFormSetHelper(FormHelper):
+    """
+    Create the crispy-forms FormHelper. The form will then be rendered
+    using {% crispy form %} in the template.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(AttachmentLinksFormSetHelper, self).__init__(*args, **kwargs)
+        self.layout = Layout(
+            Div(
+                HTML("""
+                     {{% if forloop.first != forloop.last and forloop.last and\
+                             forloop.counter != {MAX_AP_PER_USERAS} %}}
```

Comment:
This check `forloop.counter != {MAX_AP_PER_USERAS}` does not actually distinguish between an existing or a new link -- this behaves rather weirdly, as the 5th entry will always appear un-collapsed.
Maybe something can be done by looking at `formset_form` instead?
Btw. the condition can be simplified to `if not forloop.first and forloop.last ...`. Also, the `forloop.last` problem can be circumvented by using `forloop.counter == forloop.len_values`, if your PR is not accepted.

Finally, I wonder whether we should not just circumvent this problem entirely and programmatically (e.g. in the init of the AttachmentLinkForm) create a per form layout and render the formset as described here: https://django-crispy-forms.readthedocs.io/en/latest/crispy_tag_formsets.html#formset-forms-with-different-layouts.
This would also help with the hide the active and DELETE field for new entries etc.